### ISSUE TITLE
Toggling "Use Regex" clears selection in the editor

### ISSUE THREAD

**Issue by WalterGR** [atom/find-and-replace#631]
Atom 1.3.2 on OS X 10.10.5.

Repro:

1.  Select text in which you want to do a regex find-and-replace.
2.  Notice that "Use Regex" is on or off, and you want it in the other state.
3.  Toggle it, by clicking its button or using the keyboard shortcut.

Expected:

"Use Regex" is toggled and the selection in the editor is unchanged.

Actual:

"Use Regex" is toggled and the selection in the editor is cleared.

**Comment by benogle** [comment]
Selection in the editor? Does the selection match the regex pattern?

**Comment by WalterGR** [comment]
> Selection in the editor?

Yes.  I've edited the title and description to make this explicit.

> Does the selection match the regex pattern?

Do you mean "Does the selection have one or more matches of the regex pattern?" or "Is the selection in its entirety matched by the regex pattern?"

**Comment by benogle** [comment]
Can you post a gif? Can you paste in the find-and-replace section in your config?

**Comment by WalterGR** [comment]
> Can you post a gif?

I will do so shortly.

> Can you paste in the find-and-replace section in your config?

There is no such section in my config.

**Comment by lee-dohm** [comment]
I cannot reproduce this using Atom v1.3.3 on Mac OS X 10.11.2.

Could you give the following information?

* Which version of Atom are you using?
* What OS and version are you running?
* Do you have [Emmet](https://atom.io/packages/emmet) installed?
* Can you reproduce the issue after completely exiting Atom and restarting using `atom --safe`?

**Comment by WalterGR** [comment]
> Could you give the following information?  Which version of Atom are you using?  What OS and version are you running?

Apologies for not including that in the initial report.  (Added.)  I was using Atom 1.3.2.

> I cannot reproduce this using Atom v1.3.3 on Mac OS X 10.11.2.

I can confirm that it no longer repros in 1.3.3.  Yay!  :)

**Comment by WalterGR** [comment]
I was mistaken.  This still repros in 1.4.0.

I'm happy to provide a screen recording if the repro steps I've given aren't working for you.

**Comment by lee-dohm** [comment]
I believe this is expected behavior. You didn't specify what kind of document to open and whether it should or shouldn't have a `^` character pre-existing in it. So I improvised. Here's what I did:

1. Launch Atom v1.4.0 on a recent revision of the atom/atom repo
1. Open `README.md`
1. Use <kbd>Cmd+F</kbd> to open the Find/Replace pane
1. Enusure that both "Only in Selection" and "Use Regex" are off
1. Enter `.` as the Find text
1. Press <kbd>Cmd+G</kbd> to execute "Find Next"
1. Select several lines of text in the editor
1. Press <kbd>Cmd+F</kbd> to *switch focus to the already open Find/Replace pane* (this was also something that wasn't clear in your repro steps)
1. Click the "Use Regex" button

**Expected and Actual:** Because any time the state of the Find/Replace panel changes (so long as certain minimums are met) the search is re-executed, this causes the first valid search result to become selected ... which by its nature changes the selection

When using `^` as the search text, it found 82 "beginning of line" matches and even though the selection isn't visible, the selection changes just as it did in my `.` example above.

**Comment by WalterGR** [comment]
> You didn't specify what kind of document to open and whether it should or shouldn't have a ^ character pre-existing in it.

It shouldn't matter so I left it unspecified.

> Press Cmd+F to _switch focus to the already open Find/Replace pane_ (this was also something that wasn't clear in your repro steps)

I think you missed step 5: "Dismiss Find/Replace pane e.g. by pressing Esc."

> Because any time the state of the Find/Replace panel changes (so long as certain minimums are met) the search is re-executed

I'm guessing that "execute" == `find-and-replace:find-next`[1] in this context?

The thing is, when you open the Find and Replace pane and start typing, no search is executed until you manually invoke `find-next`, e.g. by pressing Cmd-g.  So it's not really _re_-executing the search, it's more: if you change search options, then maybe possibly it will execute the search for the first time, instead of waiting for you to press Cmd-g.

Find and Replace seems unnecessarily finicky in Atom.  You have to do everything in the exact right order, or you've got to navigate back to where you were in the document and re-select the text you had selected because it decided to go ahead and execute the search.

Typed text updates the boxed matches and the match count and waits for you to press Cmd-g.  The options work _completely differently_.

If this is indeed the design...  I'm sorry to hear that.

[1] - And `replace-next` and `replace-all`, but that's out of scope for this ticket...

**Comment by lee-dohm** [comment]
Because I wasn't sure what all the permutations of the expected behavior is, I'm following up with the developers on this. We'll update when we have more information.

**Comment by WalterGR** [comment]
Thanks!

**Comment by WalterGR** [comment]
My more detailed repro steps are unnecessary.  The original repro steps are fine, but to ensure you're starting from a clean slate, add:

Step 0.  Open Atom.  Open Find and Replace pane.  Ensure all options are toggled off.  Close Atom.  Open Atom.
